Chore/add a tree structure to activity and test performance

tahb opened the pull request:
## Changes in this PR

- refactor the parent relationship of activities so that instead of `activity.activity_id` we have `activity.parent_id`. This follows a found convention for Rails gems that deal trees and enables the installation of the `acts_as_tree` with minimal impact
- refactor `activity.parent_activities` by instead telling the `acts_as_tree` gem. This gem allows us find out about ancestors and children for any given node without us having to have checks on the level of each activitiy to infer how many parents might exist - quite a fragile check. It also makes it easy to extend our hierarchy by adding or removing levels in the future.
- preload activity parents when querying collections of activities (unless it's a fund as there's no current parent)

As mentioned in the commit that added [acts_as_tree](https://github.com/amerine/acts_as_tree). I investigated Ancestry and Acts_as_nested as two alternative options first. Ready the commit for more info on why I ended up picking acts_as_tree.

## Screenshots of UI changes

I created a temporary view to demonstrate what the impact of maintaining a tree structure. For 1945 activities of all levels it took the browser 2.02 seconds in total to load - this reduced to 1.3 seconds after replacing the lost parent_id database index.

The code looked like this:

```
%table.govuk-table
  %thead.govuk-table__head
    %tr.govuk-table__row
      %th.govuk-table__header
        Title
      %th.govuk-table__header
        Fund
      %th.govuk-table__header
        Programme
      %th.govuk-table__header
        Project

  %tbody.govuk-table__body
    - @activities.each do |activity|
      %tr.govuk-table__row
        %td.govuk-table__cell= activity.title
        %td.govuk-table__cell= activity.parent_activities.first&.title
        %td.govuk-table__cell= activity.parent_activities.second&.title
        %td.govuk-table__cell= activity.parent_activities.third&.title

```

I think the case can be made that a page with 100+ activities all rendered at once is not usable, and that pagination or downloading to a spreadsheet would be more appropriate. I thought it was a useful exercise to see how performant the view would be in an extreme case and I think 2 seconds is a lot better than I feared.

![Screenshot 2020-06-26 at 15 17 52](https://user-images.githubusercontent.com/912473/85867026-393b1a80-b7c0-11ea-8054-70e7457717a7.png)

Rendered the same amount of content on `develop` without these changes the time is 5.3 seconds, up from 2 seconds. 

![Screenshot 2020-06-26 at 15 22 23](https://user-images.githubusercontent.com/912473/85867558-fb8ac180-b7c0-11ea-992a-ba2a541b3ece.png)

![Screenshot_2020-06-26 — Report your official development assistance](https://user-images.githubusercontent.com/912473/85866289-12c8af80-b7bf-11ea-8be0-272f644c95fd.png)

(there were sub-projects in this page but the screenshot cropped it :) )

### 

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [x] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

tahb commented on db/schema.rb:
```diff
@@ -49,7 +48,7 @@
     t.string "sector_category"
     t.boolean "ingested", default: false
     t.string "legacy_iati_xml"
-    t.index ["activity_id"], name: "index_activities_on_activity_id"
```
Do I need to re add this back?

tahb replied:
Adding this back in for `parent_id` reduced the page load time from 2 seconds to 1.3 seconds.

mec commented on spec/support/bullet.rb:
```diff
@@ -3,6 +3,7 @@
     config.before(:each) do
       Bullet.add_whitelist type: :unused_eager_loading, class_name: "User", association: :organisation
       Bullet.add_whitelist type: :unused_eager_loading, class_name: "Activity", association: :organisation
+      Bullet.add_whitelist type: :unused_eager_loading, class_name: "Activity", association: :parent
```
This makes me a bit nervous - mostly because I don't fully understand bullet!

Can we disable bullet around the tests we know it is a false positive rather than for the whole `Activity` class?

tahb replied:
It makes me nervous too because it might hide alerts about other places we are eager loading but not using it, which may cause things to slow down unnecessarily. 

I'll take another look and see what an alternative might look like and report back here 👍

mec replied:
I `think` we could diasble bullet around the specs: https://github.com/flyerhzm/bullet#run-in-tests

mec commented on db/migrate/20200626100128_change_activity_parent_relationship_name.rb:
```diff
@@ -0,0 +1,23 @@
+    add_column :activities, :parent_id, :string
+    Activity.transaction do
+      Activity.all.each do |activity|
+        activity.update!(parent_id: activity.activity_id)
```
We shold handle the case where an activity is incomplete - as discussed setting `validate: false` is a nice clear wat to do that.

tahb commented:
@mec I've now:

1. updated the migration with `save!(validate: false)`
2. configured Bullet to only whitelist `unused_eager_loading` in the 3 service classes for `Activity.parent`. It should flag for anything else, i.e `Activity.includes(:children)`, now or in the future. I've tested that this config overwride isn't persisted to the rest of the test suite
3. I've added some more tests into each service class for the addition of `includes` for good measure
4. I noticed a second place Bullet whitelists existed and settled for merging the 2 lists together for now

Hopefully this is ready for another look!

mec commented on db/schema.rb:
```diff
@@ -162,7 +162,6 @@
     t.index ["role"], name: "index_users_on_role"
   end
 
-  add_foreign_key "activities", "activities"
```
I am trying to understand why there are no foreign keys here and what the relationship is we want, I think it's about making sure the parent of a child is not deleted, orphaning the child  but I cannot see that reflected in the schema?

tahb replied:
I missed this removal and am adding it back in. I believe it is about making sure the parent of a child is not deleted.

mec commented on spec/services/find_programme_activities_spec.rb:
```diff
@@ -8,7 +8,20 @@
   let!(:other_programme) { create(:programme_activity) }
 
+  before(:each) do
+    Bullet.add_whitelist type: :unused_eager_loading, class_name: "Activity", association: :parent
```
I've said before I am not super clear on bullet but I feel like this is just being set globally, how do we check it is removed after `each`? When I look at `Bullet.whitelist` 'after(:each)` and `after(:all)` I still see `:parent` which I think is this config?

tahb replied:
Sorry, I should have made myself more clear as I had left some new thoughts on this in the commit message:

> I tested that setting the whitelist in this way does not overwrite the global Bullet configuration. I ran 2 tests, the former with the whitelist change, and the latter without. The first file passes and the second fails. This confirms that the whitelist did not persist to other subsequant tests.

I'm surprised that the `Bullet.whitelist` is different in other tests.

tahb replied:
I think what we do here is tied to where whitelist config here which is the subject of the other comment thread here https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/494#pullrequestreview-441483206

mec commented:
@tahb I would love to understand the two comments I've left.

mec commented on config/environments/test.rb:
```diff
@@ -59,8 +59,6 @@
     Bullet.enable = true
     Bullet.bullet_logger = true
     Bullet.raise = true # raise an error if n+1 query occurs
-    Bullet.add_whitelist type: :unused_eager_loading, class_name: "Transaction", association: :provider
```
I am wondering if this is actaully the 'correct' place to add whitelists as putting them in before and after hooks mean we keep adding the same thing over and over?

tahb replied:
I take your point. I suspect the 2 points may be connected. If the config is reapplied after each test, that could be why I'm able to overwride it in my own before block that's run later. I'll check how much overhead this move removes from to the test suite run time.

tahb replied:
Yes, with my version it looks like the Bullet whitelist hash is reset after every test. 

Moving these lines to test.rb does not appear to save any noticeable time to the running of the test suite, both at 1 minute 54 seconds ish on my machine. I guess this makes sense as it is building a small hash that doesn't do any further calculations.

Their documentation does not set out a convention for us to follow on where whitelists should be configured nor trade offs to consider. Moving whitelists that do not change to initialisation rather than `before(:each)` does seem reasonable. I think the trade off is we lose the ability to control these whitelists on a per test basis - something we need for our service classes in this PR.

If we move the whitelists to test.rb I think the only alternative is to turn Bullet off for the entire service class test before, and re-enable it afterwards. This also feels like a worse choice since we would have no monitoring at all.

I'd be in favour of keeping as much bullet monitoring as we can by leaving this in bullet.rb unless there's a way I've not seen to avoid doing `Before … bullet.enable = false`, `After … Bullet.enable = true`?

tahb replied:
@leeky I'm concious that @mec may be heads down in other work. Do you have an opinion on our best course of action here?

mec replied:
@tahb here is a quick thing you can do to demonstate what I am seeing that I would lke to understand:

In `spec/support/bullet.rb` add a line after 13:

`pp Bullet.whitelist`

So we print out the current whitelist after each spec runs, now run the suite and watch that whitelist grow! 

Perhaps it doesn't matter that the whitelist is growing exponentially or I am misunderstanding, if so I would like to understand.

We are using the before hook for adding the `:parent` association on Activity in three spec files so I believe the hooks are scoped to the specs inside that file. That means we see three entries in the whitelist for `:parent` three times (three for each file). Also notice once the three specs have run the three `:parent` entries remain in the whitelist for the rest of the suite, which again points to the global nature of using `add_to_whitelist`.

Rpsec itself is configured to add four entries to the whitelist for every spec in the suite, so there a lot of entries in the whitelist. This is what makes me want to put the existing `Bullet.add_whitelist` in one place rather than in every spec.

Ultimately, this is how bullet has been setup so I am not trying to block this - I would just like to learn and understand, if we want to just move this forward, I think it is fine in a wider context!

tahb replied:
I can now see the whitelist growing linearly, though I'm not sure of the immediate impact it clearly isn't a healthy trait. 

I had wondered which set of problems we'd prefer: sensible whitelist configuration or making sure Bullet is monitoring for future performance issues in our classes. 

I've found another way which should give us both with `Bullet.delete_whitelist`. Unfortunantley I didn't spot this sooner!

```
  before(:all) do
    Bullet.add_whitelist type: :unused_eager_loading, class_name: "Activity", association: :parent
  end

  after(:all) do
    Bullet.delete_whitelist(type: :unused_eager_loading, class_name: "Activity", association: :parent)
  end
```

leeky commented:
@tahb For consistency, wouldn't it better to rename `activity_id` to `parent_activity_id`, as this name is in use in `Budget`, `Transaction` and `PlannedDisembursement`? 

Then you can drop the `Activity#parent_activity` method altogether, and just use that association as-is?

tahb commented:
@leeky `activity_id` is being renamed to`parent_id` as it appears to be a conventional column name that all the tree gems use. It feels right to keep with the convention and `parent` and `children` makes sense to me when associating objects of the same type up and down a tree.

I did notice the `parent_activity_id` column on Budgets etc. I felt that the association between Activity->Budget is not tree shaped as Activity->Activity is, and naming the columns the same may cause confusion. Arguable Budget could be renamed to have an `activity_id`, the fact it has to also include `parent` is a plaster because of a naming clash with the PublicActivity gem.

tahb commented on CHANGELOG.md:
```diff
@@ -186,13 +186,14 @@
 - Remove 3 unwanted activities from production
 - Content changes to fields for transaction value and activity identifier
 - Increase the width of the application layout
-- Activity show content is show in tabs for financials and details
```
I noticed this was in the wrong place when I rebased. Hoping it's agreeable to fix this as I go on the basis that if another PR beats this change to it, I'll have to address a conflict before this can be merged in.

tahb commented on config/environments/test.rb:
```diff
@@ -59,6 +59,8 @@
     Bullet.enable = true
     Bullet.bullet_logger = true
     Bullet.raise = true # raise an error if n+1 query occurs
+    Bullet.add_whitelist type: :unused_eager_loading, class_name: "User", association: :organisation
```
Existing whitelists now all live in test.rb.

tahb commented on spec/services/find_third_party_project_activities_spec.rb:
```diff
@@ -8,7 +8,24 @@
+    Bullet.add_whitelist type: :unused_eager_loading, class_name: "Activity", association: :parent
+  end
+
+  after(:all) do
```
We target the specific whitelist we added once we're done with all tests in the file, and remove it.

tahb commented:
@mec thank you for prodding to make this better 👍
